File: app/src/main/java/com/example/vamshi/homwayprojectchallenge/Screens/MainScreen/MainPresenter.java
package com.example.vamshi.homwayprojectchallenge.Screens.MainScreen;

import android.content.SharedPreferences;
import android.util.Log;

import com.example.vamshi.homwayprojectchallenge.Dagger.Qualifiers.QFourSquareGetRequest;
import com.example.vamshi.homwayprojectchallenge.Model.CallObserver;
import com.example.vamshi.homwayprojectchallenge.Model.JsonDistanceMatrix;
import com.example.vamshi.homwayprojectchallenge.Model.JsonOutput;
import com.example.vamshi.homwayprojectchallenge.Model.Retrofit.QueryConstants;
import com.example.vamshi.homwayprojectchallenge.Model.Retrofit.RemoteRx;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.Observer;
import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Response;

public class MainPresenter implements ContractPresenterView.PresenterMainWork {


    SharedPreferences favourites;
    RemoteRx callClient;


    @Inject
    public MainPresenter(@QFourSquareGetRequest RemoteRx callClient,
                         SharedPreferences favourites) {
        this.favourites = favourites;
        this.callClient = callClient;
    }

    @Override
    public void getPlaces(String query) {
        Observable<Response<JsonOutput>> observable = callClient.QueryPlaceDetails(
                QueryConstants.NEAR,
                QueryConstants.MATCH,
                QueryConstants.RADIUS,
                query,
                QueryConstants.FOURSQUARE_CLIENT_KEY,
                QueryConstants.FOURSQUARE_CLIENT_SECRET,
                QueryConstants.YYYMMDD);



        observable.subscribeOn(Schedulers.computation())
                .observeOn(Schedulers.computation())
                .subscribe(new CallObserver());
    }

    @Override
    public void addFavourite(String key) {

        SharedPreferences.Editor spEdit = favourites.edit();
        spEdit.putBoolean(key, true);
        spEdit.commit();

    }

    @Override
    public void removeFavourite(String key) {
        SharedPreferences.Editor spEdit = favourites.edit();
        spEdit.remove(key);
        spEdit.commit();

    }

    @Override
    public Boolean isFavourite(String key) {

        if (favourites == null) {
            return false;
        } else {
            return favourites.contains(key);
        }
    }

    @Override
    public void maStart() {


    }

    @Override
    public void maStop() {

    }


}

File: app/src/main/java/com/example/vamshi/Curious/Screens/MainScreen/MainPresenter.kt
package com.example.vamshi.Curious.Screens.MainScreen

import android.content.SharedPreferences
import android.location.Location

import com.example.vamshi.Curious.Dagger.Qualifiers.QFourSquareGetRequest
import com.example.vamshi.Curious.Model.JsonOutput
import com.example.vamshi.Curious.Model.Retrofit.QueryConstants
import com.example.vamshi.Curious.Model.Retrofit.RemoteRx
import com.google.android.gms.maps.model.LatLng

import javax.inject.Inject

import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import retrofit2.Response
import kotlin.math.truncate

class MainPresenter @Inject
constructor(@param:QFourSquareGetRequest internal var callClient: RemoteRx,
            internal var favourites: SharedPreferences?) : MainContractor.PresenterMainWork {
    internal var rxdisposable: Disposable? = null
    private var mlocation: String = "47.6,-122.3"

    override fun getPlaces(query: String): Boolean {

        val observable = callClient.QueryPlaceDetails(
                mlocation,
                QueryConstants.INTENT,
                QueryConstants.limit,
                QueryConstants.RADIUS,
                query,
                QueryConstants.FOURSQUARE_CLIENT_KEY,
                QueryConstants.FOURSQUARE_CLIENT_SECRET,
                QueryConstants.YYYMMDD)


        observable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(CallObserver())

        return true
    }


    override fun currentLocation(location: Location?) {
        var lat:Double = truncate(location?.latitude ?: 38.89)
        var lng:Double = truncate(location?.longitude ?: 77.03)

        QueryConstants.CURRENT_PLACE = LatLng(lat,lng)
        mlocation = "${lat},${lng}"
    }

    override fun addFavourite(key: String) {

        val spEdit = favourites!!.edit()
        spEdit.putBoolean(key, true)
        spEdit.commit()

    }

    override fun removeFavourite(key: String) {
        val spEdit = favourites!!.edit()
        spEdit.remove(key)
        spEdit.commit()
    }

    override fun isFavourite(key: String): Boolean {

        return if (favourites == null) {
            false
        } else {
            favourites!!.contains(key)
        }
    }

    override fun maStart() {


    }

    override fun maStop() {

    }


}
